### PR TITLE
Work with python invoke-powered build process

### PR DESCRIPTION
**Pull request by caboteria**
main_test.go still assumed the old make.go process.  This doesn't get us to working e2e tests but it's a step in the right direction.

### CODE REVIEW

**Comment by russellb** [comment]
Thanks for looking at the e2etests situation!  change lgtm

cc @rata

**Comment by caboteria** [comment]
Heh, reasonable question!  Because there's no longer a make.go file, the current main version (c65516fc0) crashes immediately like:
```
~/try/metallb$ go test ./e2etest/
stat make.go: no such file or directory
Failed to build MetalLB images: exit status 1FAIL	go.universe.tf/metallb/e2etest	0.008s
FAIL
```
With my changes the e2etests run, but they don't pass.  So this change is _necessary_ but not _sufficient_.  With this change we can work on fixing the e2etests (which are currently failing with a VDE error).
```
~/try/metallb$ go test ./e2etest/
go.universe.tf/metallb/internal/config
go.universe.tf/metallb/internal/allocator
go.universe.tf/metallb/controller
go build -v -o build/amd64/controller/controller -ldflags '-X go.universe.tf/metallb/internal/version.gitCommit=6589bb4e -X go.universe.tf/metallb/internal/version.gitBranch=test-tweaks' go.universe.tf/metallb/controller
docker build -t metallb/controller:e2e-amd64 -f controller/Dockerfile build/amd64/controller
Sending build context to Docker daemon  33.91MB
Step 1/3 : FROM alpine:latest
 ---> a24bb4013296
Step 2/3 : ADD controller /controller
 ---> 08f388714760
Step 3/3 : ENTRYPOINT ["/controller"]
 ---> Running in 6254d60bb485
Removing intermediate container 6254d60bb485
 ---> 7a6393b35710
Successfully built 7a6393b35710
Successfully tagged metallb/controller:e2e-amd64
go.universe.tf/metallb/mirror-server
go build -v -o build/amd64/mirror-server/mirror-server -ldflags '-X go.universe.tf/metallb/internal/version.gitCommit=6589bb4e -X go.universe.tf/metallb/internal/version.gitBranch=test-tweaks' go.universe.tf/metallb/mirror-server
docker build -t metallb/mirror-server:e2e-amd64 -f mirror-server/Dockerfile build/amd64/mirror-server
Sending build context to Docker daemon  2.971MB
Step 1/3 : FROM alpine:latest
 ---> a24bb4013296
Step 2/3 : ADD mirror-server /mirror
 ---> b3ad5890359f
Step 3/3 : ENTRYPOINT ["/mirror"]
 ---> Running in e95103e4e04e
Removing intermediate container e95103e4e04e
 ---> a38e175ad8ca
Successfully built a38e175ad8ca
Successfully tagged metallb/mirror-server:e2e-amd64
go: downloading github.com/hashicorp/memberlist v0.1.7
go: downloading github.com/mdlayher/ndp v0.0.0-20190419144644-012988d57f9a
go: downloading github.com/mdlayher/arp v0.0.0-20190313224443-98a83c8a2717
go: downloading github.com/mdlayher/ethernet v0.0.0-20190606142754-0394541c37b7
go: downloading github.com/mdlayher/raw v0.0.0-20190606142536-fef19f00fc18
go: downloading gitlab.com/golang-commonmark/puny v0.0.0-20180912090636-2cd490539afe
go: downloading github.com/hashicorp/go-multierror v1.0.0
go: downloading github.com/armon/go-metrics v0.0.0-20180917152333-f0300d1749da
go: downloading github.com/sean-/seed v0.0.0-20170313163322-e2103e2c3529
go: downloading github.com/hashicorp/go-msgpack v0.5.3
go: downloading github.com/miekg/dns v1.1.26
go: downloading github.com/google/btree v0.0.0-20180813153112-4030bb1f1f0c
go: downloading github.com/hashicorp/go-sockaddr v1.0.0
go: downloading github.com/hashicorp/errwrap v1.0.0
go: downloading github.com/hashicorp/go-immutable-radix v1.0.0
go.universe.tf/metallb/internal/bgp
github.com/mdlayher/ethernet
github.com/mdlayher/raw
golang.org/x/net/icmp
github.com/mdlayher/arp
github.com/mdlayher/ndp
go.universe.tf/metallb/internal/layer2
go.universe.tf/metallb/speaker
go build -v -o build/amd64/speaker/speaker -ldflags '-X go.universe.tf/metallb/internal/version.gitCommit=6589bb4e -X go.universe.tf/metallb/internal/version.gitBranch=test-tweaks' go.universe.tf/metallb/speaker
docker build -t metallb/speaker:e2e-amd64 -f speaker/Dockerfile build/amd64/speaker
Sending build context to Docker daemon  38.35MB
Step 1/3 : FROM alpine:latest
 ---> a24bb4013296
Step 2/3 : ADD speaker /speaker
 ---> 5793eec59253
Step 3/3 : ENTRYPOINT ["/speaker"]
 ---> Running in 97356b292854
Removing intermediate container 97356b292854
 ---> c675128ea299
Successfully built c675128ea299
Successfully tagged metallb/speaker:e2e-amd64
--- FAIL: TestBGP (0.00s)
    --- FAIL: TestBGP/calico (0.00s)
        vcluster.go:55: Opening universe at snapshot "calico": reading universe config: open /home/tobyc/try/metallb/e2etest-cached-universe/config.json: no such file or directory
    --- FAIL: TestBGP/flannel (0.00s)
        vcluster.go:55: Opening universe at snapshot "flannel": reading universe config: open /home/tobyc/try/metallb/e2etest-cached-universe/config.json: no such file or directory
    --- FAIL: TestBGP/weave (0.00s)
        vcluster.go:55: Opening universe at snapshot "weave": reading universe config: open /home/tobyc/try/metallb/e2etest-cached-universe/config.json: no such file or directory
FAIL
FAIL	go.universe.tf/metallb/e2etest	8.675s
FAIL
```

**Comment by caboteria** [comment]
While looking at the e2etests I noticed that the installation manifests were so old that they didn't work anymore.  I updated them which will be useful for anyone that wants to get the e2etests working with either kind or virtuakube.

**Comment by russellb** [comment]
I think the first commit should be safe to merge right away, even if the tests don't work all the way.  It's clearly better than where it was before!  

I might split the manifest changes out into another PR unless you want to just keep this PR open as the place to keep all of the changes up to getting the tests working again.  For those - I haven't looked at them closely but I wonder if there's anything custom about them?  Or could we just download these from their source instead of keeping a copy that we know will go stale?

**Comment by caboteria** [comment]
Thanks for taking a look, I'll back the manifest changes out for now

**Comment by russellb** [comment]
CC @rata

**Comment by caboteria** [comment]
ping @rata

**Comment by johananl** [comment]
Thanks @caboteria! Definitely looks like it's "more working" than the previous state :-)

I'm wondering if calling Invoke from Go code makes sense though. I feel like it might be a little weird since we're using Invoke as a build automation utility (instead of e.g. Make), which means `inv` is usually the top-level process which initiates any development-related task. Following this change we would have a Go process which spawns an Invoke process which in turn spawns Go processes...

I don't have an alternative in mind at the moment, just wanted to be clear about our intention here.

**Comment by caboteria** [comment]
Thanks for the insight!  We're spending a lot of time discussing a small, obvious improvement.  Any chance of getting it merged?

**Comment by rata** [comment]
I don't have a strong opinion on this. IIUC, it is one step in the direction of fixing virtuakube, but doesn't either completely fix it nor we are sure we want to use virtuakube.

@russellb @caboteria whenever you align regarding e2e, let me know if you think we are going on the virtuakube route in the shorterm and we can merge :)

**Comment by acnodal-tc** [comment]
Would it be better to remove the e2etest directory completely?  At least then people won't waste their time looking at it.

**Comment by caboteria** [comment]
There's a bug today in metallb/e2etest/main_test.go and this PR fixes that bug.  The discussion of how we want the e2etests to evolve is out of scope.  It's simple: if you apply this PR there will be one fewer bug in the MetalLB code base.

**Comment by russellb** [comment]
> There's a bug today in metallb/e2etest/main_test.go and this PR fixes that bug. The discussion of how we want the e2etests to evolve is out of scope. It's simple: if you apply this PR there will be one fewer bug in the MetalLB code base.

That's one view, and is why I suggested merging it earlier in this discussion.

Trying to understand and articulate the alternate viewpoint, @rata is this right? -- "The e2e test setup is unusable.  It's still unusable with just this merged, so we should just wait until there's a more complete and usable set of changes worked out."  Is that an accurate representation?

My instinct was to just merge it, but I can understand and respect that viewpoint, as well.  In that case, we should either:

* Mark this PR as `[WIP]` or Draft to indicate there's more work to do.
* Close the PR and just track ongoing "Get e2e tests working" in an issue

Either of these is OK with me and would more accurately reflect the current state.

**Comment by rata** [comment]
@russellb yeah, that's an accurate representation. Thanks!

I understood from the community meeting you two will coordinate regarding e2e tests. So, my assumption was that you will meet and decide either one of the two options (move forward with virtuakube, pursue something else, etc.).

So, really, @caboteria @russellb when you coordinate the path forward, just let me know how you want to proceed :)

Thanks!

**Comment by caboteria** [comment]
Ah, now I understand the misconception!

@rata, this bug fix isn't about "the direction of fixing virtuakube", it's a prerequisite for getting *any* e2etests working.  The code crashes way before virtuakube gets launched.  It's really about finishing commit edcced2096d684c2b2dc78c893d185dd2fd21ecd in which danderson changed the build system from a go-powered build system to a python-powered build system, but missed the e2etests.

So in short this fix is related not to virtuakube but to the python-powered build system, so it's necessary unless we plan to change how metallb is built.

Thanks, @russellb for helping to uncover this misunderstanding!

**Comment by caboteria** [comment]
Hi @rata, just checking in.  As I pointed out earlier this fix is independent of virtuakube and it's got @russellb 's "lgtm" so it would be great if you could merge it.  Thanks!

**Comment by johananl** [comment]
https://github.com/metallb/metallb/pull/639#issuecomment-656076943

As far as I'm concerned this ^ is the only thing left to figure out. Does it make sense to do Go > Invoke -> Go?

Something which could make more sense to me is to execute `inv` as the top-level command for triggering the E2E tests. Invoke will in turn build the images and then run `go test ./e2etest`.

Thoughts?

**Comment by caboteria** [comment]
I think we should _fix_ it first, then we can think about how we want to _refactor_ it.
